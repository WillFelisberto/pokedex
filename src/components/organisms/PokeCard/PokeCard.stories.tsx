import type { Meta, StoryObj } from '@storybook/react';

import { PokeCard } from './PokeCard';

const meta: Meta<typeof PokeCard> = {
  title: 'Organisms/PokeCard', // Categoria e nome do componente
  component: PokeCard, // Componente sendo documentado
  tags: ['autodocs'], // Adiciona documentação automática
  args: {
    // Valores padrão para as props
    pokemon: {
      baseStats: {
        hp: 45,
        attack: 49,
        defense: 49,
        specialAttack: 65,
        specialDefense: 65,
        speed: 45
      },
      japaneseName: 'イビス',
      region: 'Kanto',
      id: 2,
      name: 'ivysaur',
      sprites: {
        front_default:
          'https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/other/official-artwork/2.png'
      },
      types: ['grass', 'poison'],
      category: 'ivysaur',
      weight: 130,
      height: 10,
      weaknesses: ['flying', 'poison', 'bug', 'fire', 'ice'],
      evolutions: [
        {
          id: 1,
          name: 'bulbasaur',
          url: 'https://pokeapi.co/api/v2/pokemon/1/',
          sprite: 'https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/1.png',
          types: ['grass', 'poison'],
          level: null
        },
        {
          id: 2,
          name: 'ivysaur',
          url: 'https://pokeapi.co/api/v2/pokemon/2/',
          sprite: 'https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/2.png',
          types: ['grass', 'poison'],
          level: 16
        },
        {
          id: 3,
          name: 'venusaur',
          url: 'https://pokeapi.co/api/v2/pokemon/3/',
          sprite: 'https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/3.png',
          types: ['grass', 'poison'],
          level: 32
        }
      ],
      abilities: [
        { name: 'overgrow', is_hidden: false },
        { name: 'chlorophyll', is_hidden: true }
      ],
      genderRatio: {
        male: 87.5,
        female: 12.5
      }
    }
  }
};

export default meta;

// Tipo para as histórias
type Story = StoryObj<typeof PokeCard>;

// História padrão
export const Default: Story = {};

// História sem evoluções
export const SemEvolucoes: Story = {
  args: {
    pokemon: {
      baseStats: {
        hp: 45,
        attack: 49,
        defense: 49,
        specialAttack: 65,
        specialDefense: 65,
        speed: 45
      },
      japaneseName: 'イビス',
      region: 'Kanto',
      ...(meta.args?.pokemon ?? {}),
      evolutions: [], // Remove as evoluções
      id: 2, // Ensure id is defined
      name: 'ivysaur', // Ensure name is defined
      sprites: {
        front_default:
          'https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/other/official-artwork/2.png'
      },
      types: ['grass', 'poison'],
      category: 'ivysaur',
      weight: 130,
      height: 10,
      weaknesses: ['flying', 'poison', 'bug', 'fire', 'ice'],
      abilities: [
        { name: 'overgrow', is_hidden: false },
        { name: 'chlorophyll', is_hidden: true }
      ],
      genderRatio: {
        male: 87.5,
        female: 12.5
      }
    }
  }
};

// História com um Pokémon diferente
export const Charmander: Story = {
  args: {
    pokemon: {
      baseStats: {
        hp: 45,
        attack: 49,
        defense: 49,
        specialAttack: 65,
        specialDefense: 65,
        speed: 45
      },
      japaneseName: 'イビス',
      region: 'Kanto',
      id: 4,
      name: 'charmander',
      sprites: {
        front_default:
          'https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/other/official-artwork/4.png'
      },
      types: ['fire'],
      category: 'charmander',
      weight: 85,
      height: 6,
      weaknesses: ['water', 'ground', 'rock'],
      evolutions: [
        {
          id: 4,
          name: 'charmander',
          url: 'https://pokeapi.co/api/v2/pokemon/4/',
          sprite: 'https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/4.png',
          types: ['fire'],
          level: null
        },
        {
          id: 5,
          name: 'charmeleon',
          url: 'https://pokeapi.co/api/v2/pokemon/5/',
          sprite: 'https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/5.png',
          types: ['fire'],
          level: 16
        },
        {
          id: 6,
          name: 'charizard',
          url: 'https://pokeapi.co/api/v2/pokemon/6/',
          sprite: 'https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/6.png',
          types: ['fire', 'flying'],
          level: 36
        }
      ],
      abilities: [
        { name: 'blaze', is_hidden: false },
        { name: 'solar-power', is_hidden: true }
      ],
      genderRatio: {
        male: 87.5,
        female: 12.5
      }
    }
  }
};
