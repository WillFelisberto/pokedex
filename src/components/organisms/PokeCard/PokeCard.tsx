import Link from 'next/link';
import React from 'react';

import { PokeBadge } from '@/components/atoms/PokeBadge';
import { PokemonHeader } from '@/components/atoms/PokemonHeader';
import { PokemonStatus } from '@/components/molecules/PokemonStatus';

import { PokemonType } from '@/lib/utils/poketypes';
import { formatPokemonId } from '@/lib/utils/StatusConvert';

export interface PokemonProps {
  id: number;
  name: string;
  japaneseName?: string;
  region: string;
  sprites: {
    front_default: string;
  };
  types: PokemonType[];
  category: string;
  weight: number;
  height: number;
  weaknesses: PokemonType[];
  evolutions: {
    id: number;
    name: string;
    url: string;
    sprite: string;
    types: PokemonType[];
    level: number | null;
  }[];
  abilities: { name: string; is_hidden: boolean }[];
  genderRatio: {
    male: number;
    female: number;
  };
  baseStats: {
    hp: number;
    attack: number;
    defense: number;
    specialAttack: number;
    specialDefense: number;
    speed: number;
  };
}

export const PokeCard: React.FC<{ pokemon: PokemonProps }> = ({ pokemon }) => {
  return (
    <div
      data-testid="poke-card"
      key={pokemon.id}
      className="bg-white min-w-[360px] max-w-[360px] p-2 rounded-lg text-center flex flex-col gap-4"
    >
      <Link href={`/pokemon/${pokemon.id}`}>
        <PokemonHeader
          name={pokemon.name}
          spriteUrl={pokemon.sprites.front_default}
          types={pokemon.types}
        />
        <div className="p-2 text-left">
          <h2 className="text-black font-bold capitalize text-3xl" data-testid="pokemon-name">
            {pokemon.name}
          </h2>
          <p
            className="text-base text-gray-700"
            data-testid="pokemon-id"
          >{`NÂº${formatPokemonId(pokemon.id)}`}</p>
        </div>
        <div className="p-2 flex gap-2" data-testid="pokemon-types">
          {pokemon.types.map((type) => (
            <PokeBadge key={`${type}-${pokemon.id}`} type={type} />
          ))}
        </div>
        <div className="p-2 w-full flex gap-[20px]">
          <PokemonStatus height={pokemon.height} weight={pokemon.weight} />
        </div>
      </Link>
    </div>
  );
};
