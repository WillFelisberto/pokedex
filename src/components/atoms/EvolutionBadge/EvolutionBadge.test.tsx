import { render, screen } from '@/tests/test-utils';
import React from 'react';

import { PokemonProps } from '@/components/organisms/PokeCard';

import { EvolutionBadge } from '.';

// Mock do getPokemonTypeStyle para retornar valores fixos
jest.mock('@/lib/utils/poketypes/poketypes', () => ({
  __esModule: true,
  default: () => ({
    color: 'red',
    iconUrl: '/dummy-icon.png'
  })
}));

// Mock do PokeBadge para simplificar os testes, exibindo o tipo como texto
jest.mock('../PokeBadge', () => ({
  PokeBadge: ({ type }: { type: string }) => <span>{type}</span>
}));

describe('<EvolutionBadge />', () => {
  const evolutionProps: PokemonProps['evolutions'][number] = {
    level: 15,
    url: 'https://pokeapi.co/api/v2/evolution-chain/1/',
    id: 1,
    name: 'bulbasaur',
    sprite:
      'https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/other/official-artwork/1.png',
    types: ['grass', 'poison']
  };

  it('should render the evolution name and formatted id', () => {
    const { container } = render(<EvolutionBadge {...evolutionProps} />);

    expect(screen.getByRole('heading', { name: /bulbasaur/i })).toBeInTheDocument();
    expect(screen.getByText('NÂº001')).toBeInTheDocument();

    expect(container).toMatchSnapshot();
  });

  it('should render two images: one for the icon and one for the sprite', () => {
    render(<EvolutionBadge {...evolutionProps} />);

    const images = screen.getAllByRole('img');
    expect(images).toHaveLength(2);

    expect(screen.getByAltText('grass')).toBeInTheDocument();
    expect(screen.getByAltText('bulbasaur')).toBeInTheDocument();
  });

  it('should render a PokeBadge for each type', () => {
    render(<EvolutionBadge {...evolutionProps} />);
    expect(screen.getByText('grass')).toBeInTheDocument();
    expect(screen.getByText('poison')).toBeInTheDocument();
  });
});
