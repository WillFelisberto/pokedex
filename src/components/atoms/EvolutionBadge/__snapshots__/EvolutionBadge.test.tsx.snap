// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<EvolutionBadge /> should render the evolution name and formatted id 1`] = `
<div>
  <div
    class="flex flex-row w-full items-center gap-2 border border-[#E6E6E6]  rounded-[40px]"
  >
    <div
      class="rounded-[40px] relative w-[95px] h-[95px] flex items-center justify-center"
      style="background: red;"
    >
      <img
        alt="grass"
        height="65"
        src="/dummy-icon.png"
        style="mask-image: linear-gradient(135deg, rgba(0 0 0 / 50%), rgba(0,0,0,0)); filter: brightness(10);"
        width="65"
      />
      <div
        class="absolute inset-0 flex items-center justify-center"
      >
        <img
          alt="bulbasaur"
          height="120"
          src="https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/other/official-artwork/1.png"
          width="120"
        />
      </div>
    </div>
    <div
      class="text-left"
    >
      <h4
        class="font-bold text-base text-left text-black capitalize"
      >
        bulbasaur
      </h4>
      <span
        class="text-xs font-bold text-gray-700 text-left"
      >
        NÂº001
      </span>
      <div
        class="flex gap-2 flex-wrap"
      >
        <span>
          grass
        </span>
        <span>
          poison
        </span>
      </div>
    </div>
  </div>
</div>
`;
